Allow handlers to return user-defined error types

hawkw opened the pull request:
Currently, all endpoint handler functions must return a
`Result<T, HttpError>`, where `T` implements `HttpResponse`. This is
unfortunate, as it limits how error return values are represented in
the API. There isn't presently a mechanism for an endpoint to return
a structured error value of its own which is part of the OpenAPI spec
for that endpoint. This is discussed at length in issue #39.

This branch relaxes this requirement, and instead allows endpoint
handler functions to return `Result<T, E>`, where `E` is any type that
implements a new `HttpResponseError` trait. 

The `HttpResponseError` trait defines how to produce an error response
for an error value. This is implemented by `dropshot`'s `HttpError`
type, but it may also be implemented by user errors. Types implementing
this trait must implement `HttpResponseContent`, to determine how to
generate the response body and define its schema, and they must also
implement a method `HttpResponseError::status_code` to provide the
status code to use for the error response. This is somewhat different
from the existing `HttpCodedResponse` trait, which allows successful
responses to indicate at compile time that they will always have a
particular status code, such as 201 Created. Errors are a bit different:
we would like to be able to return any number of different error status
codes, but would still like to ensure that they represent *errors*, in
order to correctly generate an OpenAPI document where the error schemas
are returned only for error responses (see [this comment][1] for
details). As discussed [here][2], we ensure this by providing new
`ErrorStatusCode` and `ClientErrorStatusCode` types, which are newtypes
around `http::StatusCode` that only contain a 4xx or 5xx status (in the
case of `ErrorStatusCode`), or only contain a 4xx (in the case of
`ClientErrorStatusCode`). These types may be fallibly converted from an
`http::StatusCode` at runtime, but we also provide constants for
well-known 4xx and 5xx statuses, which can be used infallibly. The
`HttpResponseError::status_code` method returns an `ErrorStatusCode`
rather than a `http::StatusCode`, allowing us to ensure that error types
always have error statuses and generate a correct OpenAPI document.

Additionally, while adding `ErrorStatusCode`s, I've gone ahead and
changed the `dropshot::HttpError` type to also use it, and changed the
`HttpError::for_client_error` and `HttpError::for_status` constructors
to take a `ClientErrorStatusCode`. Although this is a breaking change,
it resolves a long-standing issue with these APIs: currently, they
assert that the provided status code is a 4xx internally, which is often
surprising to the user. Thus, this PR fixes #693.

Fixes #39
Fixes #693
Fixes #801

This branch is a second attempt at the change originally proposed in PR
#1164, so this closes #1164. This design is substantially simpler, and
*only* addresses the ability for handler functions to return
user-defined types. Other changes made in #1164, such as a way to
specify a global handler for dropshot-generated errors, and adding
headers to `HttpError` responses, can be addressed separately. For now,
all extractors and internal errors still produce `dropshot::HttpError`s.
A subsequent change will implement a mechanism for providing alternate
presentation for such errors (such as an HTML 404 page).

[1]:
    https://github.com/oxidecomputer/dropshot/issues/39#issuecomment-2483898918
[2]:
    https://github.com/oxidecomputer/dropshot/issues/39#issuecomment-2486602655

ahl commented on .envrc:
```diff
@@ -0,0 +1 @@
+use nix;
```
what's this?

hawkw replied:
agh, my bad; will back that out!

hawkw commented:
it occurs to me that `HttpResponseHeaders` might want to forward an implementation of `HttpResponseError` when the wrapped type implements that. I'll add one.

ahl commented on dropshot/src/error.rs:
```diff
@@ -86,7 +111,7 @@ pub struct HttpError {
     // by some enum type?
```
I'm not sure about the first part of this TODO, but the "should this be parametrized by some enum type" seems (effectively?) resolved by your change here. i.e. in light of custom error types I don't think we'd do that.

hawkw replied:
Yeah. Personally, I think we can probably drop the whole comment, since IMO, the
point of requiring the error code field is to encourage user code to report more
structured errors, which...we have a different solutuon for.

ahl commented on dropshot/src/error.rs:
```diff
@@ -316,6 +343,657 @@ impl Error for HttpError {
+    /// The input was not a client error (4xx) status code.
+    #[error(transparent)]
+    NotAClientError(#[from] NotAClientError),
+    /// The nput was not a valid number, was less than 100, or was greater than
```
```suggestion
    /// The input was not a valid number, was less than 100, or was greater than
```

ahl commented on dropshot/src/handler.rs:
```diff
@@ -262,7 +265,89 @@ where
+
+/// Errors returned by an endpoint handler function.
+///
+/// A function implementing an HTTP endpoint must return a `Result` with this
```
I thought it could also return whatever that implements HttpResponse?

ahl commented on dropshot/src/handler.rs:
```diff
@@ -262,7 +265,89 @@ where
+/// Errors returned by an endpoint handler function.
+///
+/// A function implementing an HTTP endpoint must return a `Result` with this
+/// error type. Typically, user code will only interact with this type via its
```
```suggestion
/// error type. Typically, consumers will only interact with this type via its
```

maybe?

ahl commented on dropshot/src/handler.rs:
```diff
@@ -262,7 +265,89 @@ where
+/// A function implementing an HTTP endpoint must return a `Result` with this
+/// error type. Typically, user code will only interact with this type via its
+/// `From` implementation, which allows it to be constructed from any value
+/// implementing the [`HttpResponseError`] trait. This type is largely an
+/// implementation detail, which is necessary to represent errors which may be
+/// errors returned by the endpoint's handler function *or* generated internally
+/// (i.e. by extractors).
```
forgive me, but this lost me

ahl commented on dropshot/src/handler.rs:
```diff
@@ -262,7 +265,89 @@ where
+///
+/// 1. Implement the [`HttpResponseContent`] trait, defining the content of the
+///    error's response body.  This trait is implemented for all types that
+///   implement the [`Serialize`] and [`JsonSchema`] traits.
```
```suggestion
///    implement the [`Serialize`] and [`JsonSchema`] traits.
```

Or perhaps: this is most simply done by implementing the `Serialize` and `JsonSchema` traits

and maybe: ...for which there's a blanket implementation

ahl commented on dropshot/src/handler.rs:
```diff
@@ -262,7 +265,89 @@ where
+/// 2. Implement [`std::fmt::Display`].  This is used in order to log an internal
+///    error message when returning an error response to the client.
+/// 3. Implement this trait's [`HttpResponseError::status_code`] method, which
+///    determines the status code for the error response.  Note that this method
```
```suggestion
///    specifies the status code for the error response.  Note that this method
```

ahl commented on dropshot/src/handler.rs:
```diff
@@ -262,7 +265,89 @@ where
+///    returns an [`ErrorStatusCode`]: a refinement of the [`http::StatusCode`]
+///    type which is constrained to only 4xx and 5xx status codes.
+///
+/// Dropshot's [`HttpError`] type also implements `HttpResponseError`, so
+/// handler functions may return `Result<T, HttpError>`.
```
```suggestion
/// [`HttpError`] type implements `HttpResponseError`, so
/// handlers may return `Result<T, HttpError>` without the need to define a custom error type.
```

maybe?

ahl commented on dropshot/src/handler.rs:
```diff
@@ -262,7 +265,89 @@ where
+///
+/// Dropshot's [`HttpError`] type also implements `HttpResponseError`, so
+/// handler functions may return `Result<T, HttpError>`.
+///
```
```suggestion
```

ahl commented on dropshot/src/handler.rs:
```diff
@@ -584,6 +669,29 @@ impl HttpResponse for Response<Body> {
     }
 }
 
+impl<T, E> HttpResponse for Result<T, E>
```
I know you've already talked about this above, but this is such an important impl that perhaps it warrants a comment?

Implement HttpResponse for any Result<HttpResponse, HttpResponseError> to address the most common use case

Now that I've written that... meh... I don't know. Do what you like!

ahl commented on dropshot/src/handler.rs:
```diff
@@ -676,11 +784,50 @@ where
 
+impl HttpResponseContent for HttpError {
+    fn to_response(self, _: http::response::Builder) -> HttpContentResult {
+        // Okay, here is where we do a devious little bit of slight-of-hand.
```
```suggestion
        // Okay, here is where we do a devious little bit of sleight-of-hand.
```

ahl commented on dropshot/src/lib.rs:
```diff
@@ -416,10 +416,18 @@
+//! A [`Result`]`<T, E>` implements `HttpResponse` when the `Ok` type (`T`)
+//! implements `HttpResponse` and the `Err` type (`E`) implements the
+//! [`HttpResponseError`] trait, so endpoint functions may be fallible.
+//! User-defined error types may implement the [`HttpResponseError`] trait in
+//! order to define how to produce an HTTP response from that error.
+//! Alternatively, Dropshot's [`HttpError`] type implements
+//! [`HttpResponseError`] and can be used as the handler's error type.
```
Endpoint functions may return the [`HttpError`] type in the error case or a user-defined error type that implements the [`HttpResponseError`] trait. [`HttpError`] may be simpler, while a custom error type permits greater expressivity in the case of an error.

ahl commented on dropshot/src/router.rs:
```diff
@@ -1,6 +1,7 @@
 // Copyright 2021 Oxide Computer Company
```
update here and elsewhere as needed

ahl commented on dropshot/src/server.rs:
```diff
@@ -943,9 +946,9 @@ async fn http_request_handle<C: ServerContext>(
                         ),
```
should we be firing a USDT probe here?

ahl replied:
Probably doesn't... (unless you've done it already). We can file an issue

ahl commented on dropshot/tests/fail/bad_endpoint10.stderr:
```diff
@@ -1,17 +1,22 @@
-   |      ^^^^^^
-   |      |
-   |      expected this to be `HttpError`
-   |      type mismatch resolving `<String as TypeEq>::This == HttpError`
+   |      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `HttpResponse` is not implemented for `Result<HttpResponseOk<()>, String>`
```
This seems like a step backward; could `#[diagnostic]` help here?

hawkw replied:
What do you think of the new error message in https://github.com/oxidecomputer/dropshot/pull/1180/commits/f9642d16f850869f98fdec2cc51528f0ba4f07cc?

ahl commented on dropshot/tests/fail/bad_endpoint12.stderr:
```diff
@@ -1,30 +1,22 @@
   --> tests/fail/bad_endpoint12.rs:15:6
    |
 15 | ) -> Result<String, HttpError> {
-   |      ^^^^^^ the trait `HttpCodedResponse` is not implemented for `String`, which is required by `Result<String, HttpError>: ResultTrait`
+   |      ^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `HttpResponse` is not implemented for `Result<String, HttpError>`
```
as above; I **think** this might be harder for people to understand? What do you think?

ahl commented on dropshot_endpoint/src/endpoint.rs:
```diff
@@ -384,34 +384,11 @@ impl<'ast> EndpointParams<'ast> {
-                    type This = Self;
-                }
-                fn validate_result_error_type<T>()
+                fn validate_response_type<T>()
```
all of this is extraneous? anything relevant here to juice up the error messages?

ahl reviewed (commented):
thanks for doing this

hawkw commented on dropshot/src/error.rs:
```diff
@@ -28,20 +28,45 @@
 //! we recommend a separate enum to represent their errors in an HTTP-agnostic
 //! way.  Consumers can provide a `From` implementation that converts these
-//! errors into HttpErrors.
+//! errors into `HttpError`s, or implement the [`HttpResponseError`] trait to
+//! provide their own mechanism
```
I think I like your wording better, thanks! I'll update this.

hawkw commented on dropshot/src/error.rs:
````diff
@@ -316,6 +343,657 @@ impl Error for HttpError {
+    /// let err = ErrorStatusCode::from_u16(99);
+    /// assert!(err.is_err());
+    /// ```
+    #[inline]
````
The `#[inline]` attributes are here just because I lifted the function signatures for these conversions straight from the `http` crate and then modified them, it inlines these conversions. I'm happy to remove them, I doubt they have much of an impact.

ahl replied:
Aha! That makes a ton of sense and I'm happy to leave them (or keep them). Whatever you prefer.

hawkw commented on dropshot/src/error.rs:
```diff
@@ -178,14 +203,16 @@ impl HttpError {
     /// `message` used for both the internal and external message.  The
     /// expectation here is that for most 400-level errors, there's no need for a
     /// separate internal message.
+    ///
+    /// The status code is provided as a [`ClientErrorStatusCode`], which may
+    /// only be constructed for 4xx client errors.
```
I felt it was potentially helpful to state explicitly that the purpose of that
type was to ensure that this function takes only a client error, but...you're
right, the name of the function  is `for_client_error`, and the name of the type
is `ClientError...`, so that should, hopefully, be pretty obvious without this.

hawkw commented on dropshot/src/api_description.rs:
```diff
@@ -919,27 +925,59 @@ impl<Context: ServerContext> ApiDescription<Context> {
                 }
             };
 
+            // If the endpoint defines an error type, emit that for
```
So, I'd like to put them in `components.responses`, too. The reason I didn't is that it might be a bit annoying to determine the name for each response schema. `schemars` internally disambiguates colliding schema names by turning subsequent ones into like `Error2` or whatever, but (AFAICT) we only get that when we actually generate the schema and it gives us back a reference (into `components.schemas`). We could then try to parse that reference and get the name back out to then use it to generate a `components.responses` entry for that response, which seems possible, I just thought it seemed annoying enough that I didn't *really* want to bother with it. Do you think it's worth doing?

hawkw replied:
I don't believe that deduplication is applied to `JsonSchema::schema_name()`; as
far as I can tell, it only happens once a schema has already been generated,
because that's when the generator can check if the name already exists in the
set of schemas that have been generated so far?

hawkw replied:
Ah. Unfortunately, it turns out that Progenitor chokes on outputting two separate response objects that have the same response body schema; it currently expects there to be exactly one response object for errors, as the current code hits [this assertion](https://github.com/oxidecomputer/progenitor/blob/96bdb54c12f5ba02b6ca5f9213083e8d9c4b1e4f/progenitor-impl/src/method.rs#L1277-L1279).

I'm on the fence about whether the right solution here is to change the OpenAPI spec generation in Dropshot to emit a single response schema object for 4xx and 5xx errors, or to make Progenitor smart enough to understand that multiple response schemas can share the same _response body type schema_, and recognize that those are the same thing. What do you think @ahl?

ahl replied:
What's the OpenAPI input it's choking on? I wouldn't have thought that this move away from `components/responses` would have had this result.

With regard to progenitor supporting it; I would like to do that, but I would also love to be able to defer that. It's not a small amount of work to do reasonably well.

hawkw replied:
Progenitor panics when building oxidecomputer/omicron#7196, which updates Omicron to depend on this branch. Note that the only change to the schemas is to generate separate response schemas for 4xx and 5xxs, as that PR hasn't added any user-defined error types or anything. It'll die on any of the OpenAPI docs in that PR, which one it hits first seems to be non-deterministic and depends on the order in which crates are built. For an example of the OpenAPI change, see https://github.com/oxidecomputer/omicron/pull/7196/files#diff-a389634108eedc06638f772e1384ef397e4d4254363b3459e5043341227f06b6.

I'm going to take another crack at having Dropshot generate one response schema for both error status codes, which would hopefully resolve the issue without requiring a change to Progenitor.

ahl replied:
I'm genuinely surprised that progenitor is dying despite its constraints. My recollection was that a reference to a response would be handled identically to an inline response. I'm going to look on the progenitor side.

hawkw replied:
Okay @ahl, if you look at the PR comparison view again as of commit f0e6a46acbcb21774cc50ba24ee2faa0b9e060a3, you'll see that there's now *no diff* in any of the generated OpenAPI documents compared to `main`; the only diff in the expectorate output now is the new schemas with user-defined error types!  

So, while it would be nice for progenitor to be able to handle cases like this, APIs that don't use custom error types will now emit backwards-compatible OpenAPI documents that the current version of Progenitor ought to be able to handle correctly.

ahl commented on dropshot/src/handler.rs:
```diff
@@ -268,17 +268,38 @@ where
+/// 3. The response returned by the handler function fails to be converted into
+///    a `Response<Body>`.
+///
+/// In cases (1) and (3), the error will always be an [`HttpError`], but in case
```
Oof! As discussed in DMs, this is kind of awful. In particular it means that sometimes clients will see the custom error type and sometimes not! It's also a mess with regard to OpenAPI. We **could** just say "the error is `oneOf` these types", and that's awful too!

As discussed, we might add conversion methods to `HttpResponseError` for cases 1 and 3 here.

ahl commented on dropshot/tests/fail/bad_trait_endpoint12.stderr:
```diff
@@ -1,28 +1,20 @@
+   |          ^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `HttpResponse` is not implemented for `Result<std::string::String, dropshot::HttpError>`
    |
-   = help: the following other types implement trait `HttpCodedResponse`:
-             HttpResponseAccepted<T>
-             HttpResponseCreated<T>
-             HttpResponseDeleted
-             HttpResponseOk<T>
-             HttpResponseUpdatedNoContent
```
These were really useful suggestions. Is there any way to get these back?

hawkw replied:
They're back as of https://github.com/oxidecomputer/dropshot/pull/1180/commits/00bcea760a2286dbf499deae550986ff2cef8fd1, since we no longer allow endpoints to return `T: HttpResponse` and instead require `Result<T, E> where T: HttpResponse, E: HttpResponseError` again. The error messages are now much closer to how they were before except that the error type can be any `E: HttpResponseError` rather than `HttpError`.

ahl commented on dropshot/tests/fail/bad_endpoint10.stderr:
```diff
@@ -1,20 +1,18 @@
    |
 16 | ) -> Result<HttpResponseOk<()>, String> {
-   |      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `HttpResponse` is not implemented for `Result<HttpResponseOk<()>, String>`
+   |      ^^^^^^ the trait `HttpResponseError` is not implemented for `String`
```
do we want a #[diagnostic] that helps users find HttpError?

ahl reviewed (commented):
I like this direction

ahl reviewed (approved):
nice work

hawkw commented on dropshot/tests/fail/bad_endpoint10.stderr:
```diff
@@ -4,6 +4,7 @@ error[E0277]: the trait bound `String: HttpResponseError` is not satisfied
 16 | ) -> Result<HttpResponseOk<()>, String> {
    |      ^^^^^^ the trait `HttpResponseError` is not implemented for `String`
    |
+   = note: consider using `dropshot::HttpError`, unless custom error presentation is needed
```
Great!

hawkw commented on dropshot/src/handler.rs:
```diff
@@ -326,7 +319,24 @@ impl HandlerError {
+                        rsp.headers_mut()
+                            .insert(crate::HEADER_REQUEST_ID, header);
+                    }
+                    // This should never happen, but let's not panic in release
+                    // mode if it does.
+                    Err(e) if cfg!(debug_assertions) => {
+                        unreachable!("request ID {request_id:?} is not a valid HeaderValue: {e}");
+                    }
```
Hmmm, the reason I didn't make this an `unwrap()` or similar is that elsewhere, when we set the request ID header when converting a  `HttpError` into a response, we do it through `http::response::Builder::header()`, which will cause `Builder::build()` to return an `Err` in the event of an invalid header value, rather than unwrapping it. See: https://docs.rs/http/latest/src/http/response.rs.html#611-624

So, perhaps what we really should do is instead return a 500, to be consistent with that behavior...but we already have a response explaining the original error, and it seemed sad to throw that away in favor of a 500 because we somehow generated a request ID that contained invalid characters.

hawkw replied:
Oh, actually, I guess that `HttpError::into_response()` also `unwrap()`s the entire `Builder::build()` result, which includes this error: https://github.com/oxidecomputer/dropshot/blob/5377f2d6de478a0e23ef27f66531c8903e5f6b72/dropshot/src/error.rs#L303

...but on the other hand, there's a comment there saying we maybe shouldn't do that:

https://github.com/oxidecomputer/dropshot/blob/5377f2d6de478a0e23ef27f66531c8903e5f6b72/dropshot/src/error.rs#L279-L286

hawkw replied:
I'll just make it panic unconditionally, I guess, although I prefer the match here rather than `unwrap`/`expect` as in this case, we know the specific header value that was invalid...

davepacheco reviewed (commented):
Very nice!  Thanks again for both taking this on and sticking with it through all the design iteration.  I don't have anything major here but a few things to fix up, plus: for all breaking changes I generally ask:

- We should update CHANGELOG.adoc. it should have super easy-to-follow instructions there for updating across this change. There should be plenty of examples -- usually we try to say how to know you're affected and exactly what code changes to make, with an example or two.
- I think it's worth doing a test-update of Omicron to a dropshot with this change. We're going to have to do it anyway and doing it first sometimes smokes out cases where the breaking change was worse than we thought to move past or stuff like that.

I did not review the bad-endpoint error output, the OpenAPI spec changes, or the dropshot_endpoint changes FWIW.

hawkw commented on dropshot/src/error.rs:
```diff
@@ -245,7 +266,7 @@ impl HttpError {
     /// Found").
     pub fn for_status(
```
yeah, good call --- while we're breaking this interface, we may as well make the name clearer, too!

hawkw commented on dropshot/src/handler.rs:
```diff
@@ -258,11 +259,282 @@ where
+/// [`HttpResponseError`].  We type-erase such errors by eagerly converting them
+/// into `Response<Body>` within the `HttpHandlerFunc` implementation, to permit
+/// the API to consist of handlers with any number of different error types.
+pub enum HandlerError {
```
It is not, and it should not be.

hawkw commented on dropshot/src/error.rs:
```diff
@@ -316,6 +337,649 @@ impl Error for HttpError {
+}
+
+macro_rules! impl_status_code_wrapper {
+    (impl StatusCode for $T:ident {
```
This is just me being cutesy about the input syntax to the macro; it could just be `$T:ident,  $NotAnError:ident,  $Invalid:ident` instead, but I thought it was fun to make it look like an actual `impl` block. It does also give us a nice way to see which type is which when reading the code. But, if this is confusing, I could get rid of it.

hawkw commented on dropshot/src/error.rs:
```diff
@@ -316,6 +337,649 @@ impl Error for HttpError {
+            }
+        }
+
+        impl PartialEq<u16> for $T {
```
The `Eq`/`PartialEq` implementations that compare the type with _other_ types can be derived (at least, without some crate that does that). We are already deriving the `PartialEq<Self>` and `Eq<Self>` implementations: https://github.com/oxidecomputer/dropshot/blob/9d99131ccd47ad138f555ed9037fef1ea5f2146a/dropshot/src/error.rs#L354

Regarding `Deref`, we _could_ do that...I feel like there's sort of an attitude that using `Deref` in that way for, essentially, implementing a kind of inheritance is bad style and that `Deref` should only be used for things that are actually pointers. But, I don't care that much about this particular style point personally, I was just following it because I figured others might. However, implementing `Deref` does mean that methods like `is_informational`, `is_success` , and `is_redirection` can also be called on an `ErrorStatusCode`. Stopping people from calling those methods on this type that isn't _that_ important, since they're not going to cause real problems --- they'll just always return `false`. But it is kind of nonsensical, so not exposing them seems good.

IMO just copying and pasting the method signatures is conceptually simpler than using `Deref` for this, although it involves more code. That's not a strongly-held opinion, though, so I'm very open to being convinced otherwise.

hawkw replied:
cab73b6791b9d7b2dee19ec75d84537f4f235c1a added some documentation about this --- what do you think?

hawkw commented on dropshot/src/test_util.rs:
```diff
@@ -334,10 +377,181 @@ impl ClientTestContext {
+    _error: PhantomData<fn(E)>,
+}
+
+impl<E> Clone for TypedErrorClientTestContext<'_, E> {
```
The derived `Clone` implementation for a generic type will generate `Clone` bounds for each of its type parameters. Unfortunately, it's not smart enough to reason that a `PhantomData` is `Clone` and eliminate those bounds. See, for example, [this playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0b3e303c8da6abecfe2556a96e99bd2a), which doesn't compile.

hawkw commented on dropshot/src/test_util.rs:
```diff
@@ -334,10 +377,181 @@ impl ClientTestContext {
+    /// we can use to invoke this endpoint from the client.  This essentially
+    /// appends the path to a base URL constructed from the server's IP address
+    /// and port.
+    pub fn url(&self, path: &str) -> Uri {
```
Yup, except that the error types are changed. Unfortunately that felt like the simplest way to do this without changing the `ClientTestContext` interface in ways that I felt made it unpleasant for the user; we could do something like turning these into default methods on a trait which both `ClientTestContext` and `TypedErrorClientTestContext` would implement, or something, but then the user has to import that trait...

hawkw commented on dropshot/src/server.rs:
```diff
@@ -943,9 +946,9 @@ async fn http_request_handle<C: ServerContext>(
                         Err(error) => {
                             warn!(request_log, "request completed after handler was already cancelled";
-                                "response_code" => error.status_code.as_str(),
-                                "error_message_internal" => error.external_message,
-                                "error_message_external" => error.internal_message,
```
yeah...

hawkw commented on dropshot/src/error.rs:
```diff
@@ -316,6 +337,649 @@ impl Error for HttpError {
+
+impl ErrorStatusCode {
+    error_status_code_constants! {
+        /// 400 Bad Request
+        /// [[RFC7231, Section 6.5.1](https://tools.ietf.org/html/rfc7231#section-6.5.1)]
+        BAD_REQUEST;
```
I just copied it from `http::StatusCode`. If new statuses are standardized and the `http` crate adds constants for them, we'd need to update this. The only way around that I can think up is a *really* gnarly build script that checks out the `http` source code and parses the status code constants they define, which... feels worse than doing it manually...

hawkw commented on dropshot/tests/integration-tests/custom_errors.rs:
```diff
@@ -0,0 +1,304 @@
+// Copyright 2024 Oxide Computer Company
+
+//! Test cases for user-defined error types.
```
Done in ab1d903ac6ff6fde3ebf3d0a9867f4a0ff5fc43a! I also added test cases for trait-based endpoint functions returning the wrong error type in ea5c9efae55e03e8184cdc9e27e458c6abf1b052, and then made some improvements to that error message in 5bf8aa26425e80b76b5a5912435d0629eccfc886.

hawkw commented on dropshot/tests/integration-tests/versions.rs:
```diff
@@ -467,7 +467,7 @@ fn test_versions_openapi_same_names() {
 
     assert_eq!(func_mods_v1, func_overrides_v1);
-    assert_eq!(func_mods_v1, traits_v1);
+    assert_eq!(func_mods_v1, traits_v1,);
```
fixed in 6ad3bb4daef675d38ff9d2979c83ea1f67e1c249, my bad!

hawkw commented on dropshot/src/error.rs:
```diff
@@ -316,6 +337,649 @@ impl Error for HttpError {
     }
 }
 
+/// An HTTP 4xx (client error) or 5xx (server error) status code.
```
done in 71dc879e6a8f08475cdc93042b9422280229a6e4!

hawkw commented on CHANGELOG.adoc:
```diff
@@ -26,6 +26,84 @@ Defining the old config option will produce an error, guiding you to perform the
++
+For details on how to implement `HttpResponseError` for user-defined types, see
+the trait documentation, or
+https://github.com/oxidecomputer/dropshot/blob/main/dropshot/examples/custom-error.rs[`examples/custom-error.rs`].
```
this link will be broken until this PR has merged --- it still felt like the nicest way to reference the example in the changelog, IMO.

hawkw commented:
Okay, @davepacheco, I've updated Omicron to use this branch in https://github.com/oxidecomputer/omicron/pull/7196, and uncovered an issue with the generated OpenAPI docs, which I've resolved (see https://github.com/oxidecomputer/dropshot/pull/1180/#discussion_r1868299626).  I've added changelog entries describing the breaking changes and the  migration instructions; let me know what you think of https://github.com/oxidecomputer/dropshot/blob/75cbc09542cf439412c4c13ee2b799f1dff1f28a/CHANGELOG.adoc#unreleased-changes-release-date-tbd

davepacheco reviewed (approved):
Thanks for the fixups and comments.  The CHANGELOG notes look terrific!  (I haven't re-reviewed all the diffs here.)
